pep-9999.rst

```diff
@@ -394,6 +403,88 @@ been fully uploaded.
 A successful cancelation request **MUST** response with a ``204 No Content``.
 
 
+Create a Draft Session
+~~~~~~~~~~~~~~~~~~~~~~
+
+The semantics for creating a draft session is identical to creating a regular
+session, with the following exception:
+
+- the ``draft`` field is present and set to ``true``;
+
+Similarly, the server response is identical to creating a regular session, with
+the following exception:
+
+- the ``url`` field is present and represents an absolute URL for the session
+
+Draft sessions support re-uploading a file with the same filename as a file
+from a previous draft session, but with different contents. If the server
+recieves a request to create a draft session for a filename that has been
+included in a previous draft session which hasn't yet been closed, the server
+should consider the previously uploaded file to be completed regardless of the
+status of the previous upload.
+
+Servers **MAY** reuse session URLs of previous sessions, and **SHOULD** group
+session URLs based on normalized project name and version.
+
+Draft sessions must be explicitly closed before the files are published, and
+servers **SHOULD NOT** automatically publish draft sessions when all files in
+the session have been uploaded. See 'Publishing a Draft Session' below.
+
+
+List Draft Sessions
+~~~~~~~~~~~~~~~~~~~
+
+In order to determine which draft sessions the server is aware of and are
+pending publication, servers **MUST** support enumerating draft sessions. When
```

Comment:
Why do we need to support enumeration? If you've already gotten a session, can't you just use the URL you've already received?

Reply:
I'm imagining a release process like the one used for pyca/cryptography, where they build and upload from multiple different environments (to get platform/architecture coverage). This means that a specific build environment doesn't have a priori knowledge of other related upload sessions, but you still want a way to collect all the artifacts for a given release (with multiple upload sessions) maybe run some tests or something, and then request that the sessions are all published simultaneously (which this doesn't necessarily account for since they'd have to make 1 request per upload session, but I'm assuming we'd provide the same session url for the same name+version, so it would be 1 request for PyPI)

I'm also imagining a workflow where a user iterating on multiple candidate artifacts for a given release via drafts, and doesn't have a great way to persist the upload session url between attempts (like maybe they're running `twine upload --draft` multiple times). And since we're saying the URL is not predictable and can change, we'd need a way to determine which sessions are available to publish when they run `twine publish` or whatever